edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs

```diff
@@ -0,0 +1,938 @@
+                                    if (int.TryParse(hostBinding.HostPort, out int hostPort))
+                                    {
+                                        // If a port entry contains the same "port", then remove it and replace with a new ServicePort that contains a target.
+                                        var duplicate = portList.SingleOrDefault(a => a.Port == hostPort);
+                                        if (duplicate != default(V1ServicePort))
+                                        {
+                                            portList.Remove(duplicate);
+                                        }
+
+                                        portList.Add(new V1ServicePort(hostPort, name: $"HostPort-{port}-{protocol.ToLower()}", protocol: protocol, targetPort: port));
+                                        onlyExposedPorts = false;
+                                    }
+                                    else
+                                    {
+                                        Events.PortBindingValue(module, portBinding.Key);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (portList.Count > 0)
+            {
+                // Selector: by module name and device name, also how we will label this puppy.
+                var objectMeta = new V1ObjectMeta(annotations: serviceAnnotations.GetOrElse(() => null), labels: labels, name: KubeUtils.SanitizeDNSValue(moduleIdentity.ModuleId));
+                // How we manage this service is dependent on the port mappings user asks for.
+                // If the user tells us to only use ClusterIP ports, we will always set the type to ClusterIP.
+                // If all we had were exposed ports, we will assume ClusterIP. Otherwise, we use the given value as the default service type
+                //
+                // If the user wants to expose the ClusterIPs port externally, they should manually create a service to expose it.
+                // This gives the user more control as to how they want this to work.
+                string serviceType;
+                if (onlyExposedPorts)
+                {
+                    serviceType = "ClusterIP";
+                }
+                else
+                {
+                    serviceType = this.defaultMapServiceType;
+                }
+
+                return Option.Some(new V1Service(metadata: objectMeta, spec: new V1ServiceSpec(type: serviceType, ports: portList, selector: labels)));
+            }
+            else
+            {
+                return Option.None<V1Service>();
+            }
+        }
+
+        private bool ValidateProtocol(string dockerProtocol, out string k8SProtocol)
+        {
+            bool result = true;
+            switch (dockerProtocol.ToUpper())
+            {
+                case "TCP":
```

Comment:
Can these be added to an enum

Reply:
Usually, I would agree, but with only 3 and in an isolated case, I don't see the benefit.